### PR TITLE
Trigger nnsLink event on keypress and click

### PR DESCRIPTION
**Pull request by coskucinkilic**
# Motivation

Fix canister button reload issue

# Changes

1. Canister button now calls nnsLink event onClick and keypress.
2.  LinkToCanisters po and test has been added.
3. AccountMenu is updated to have single function for closing popover.
4. $app/navigation mock has been updated so we can register callbacks and call them before actually navigating. rather than directly triggering the callback. Due to current implementation of jsdom not supporting native nagivate, mockLinkClickEvent has been introduced to support link clicks.
5. goto props has been updated to have correct keys and types
6. [Separate PR](https://github.com/dfinity/gix-components/pull/491) will be updating gix-components Popover menu and eliminate stopPropagation on onClick and keyPress events, which was the root cause of this issue.
7. vitest.setup missing imports have been added.

# Tests

New tests check the presence of Link, it's href, text, having an icon, dispatching nnsLink event

# Todos

- [ ] Add entry to changelog (if necessary).
not necesary

### CODE REVIEW

File: frontend/src/lib/components/header/LinkToCanisters.svelte [hunk 1]
```diff
@@ -2,9 +2,20 @@
...
+
+  const closeMenu = () => dispatcher("nnsLink");
 </script>
 
-<a data-tid="canisters-button" href={$canistersPathStore} class="text">
+<a
+  data-tid="canisters-button"
+  href={$canistersPathStore}
+  on:keypress={closeMenu}
+  on:click={closeMenu}
```
Comment (peterpeterparker) [inline frontend/src/lib/components/header/LinkToCanisters.svelte:16]: Isn't the use of a click and href combined generally considered as a bad practice? That's my impression so unless there is a source for it, I would avoid it.

From MDN [documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a):

> Use a <button> instead. In general, you should only use a hyperlink for navigation to a real URL.

We actually had the same issue in Oisy as the one you are trying to resolve (see this [PR](https://github.com/dfinity/oisy-wallet/pull/1991)). Ultimately we just decided to use a `button` until we figure out a solution that both works and comply with standards.

Comment (coskucinkilic) [reply frontend/src/lib/components/header/LinkToCanisters.svelte]: It would be nice to have this kind of Link component for now I just deleted keypress and onClick and use beforeNavigate to trigger the nnsLink event.

Comment (peterpeterparker) [reply frontend/src/lib/components/header/LinkToCanisters.svelte]: > beforeNavigate to trigger the nnsLink event.

And it does the job? It works out?

Comment (coskucinkilic) [reply frontend/src/lib/components/header/LinkToCanisters.svelte]: I mean it is jut the same button without onclick and keypress, it doesn't fix the reload issue but it is definitely nicer.

The main fix for reload issue is the gix-components updates, which removes the stopPropagation. That fixes the bug we have.

File: frontend/__mocks__/$app/navigation.ts [hunk 2]
```diff
@@ -37,3 +37,13 @@ export const afterNavigate = (callback: (navigation: Navigation) => void) => {
     delta: undefined,
   });
 };
+
+export const beforeNavigate = (callback: (navigation: Navigation) => void) => {
+  callback({
```
Comment (dskloetd) [inline frontend/__mocks__/$app/navigation.ts:42]: This mock behavior seems incorrect.
This calls the callback immediately, right?
But the real behavior only calls the callback if/when navigation occurs.
Wouldn't this cause the menu to close as soon as it's opened and the Canisters button is rendered?

(`afterNavigate` seems to have the same problem but I don't know where it's called.)

File: frontend/src/lib/components/header/LinkToCanisters.svelte (hunk 1 above)
Comment (peterpeterparker) [reply frontend/src/lib/components/header/LinkToCanisters.svelte]: Nice, thanks for sharing!

Looks good to me but I let David approve the PR ultimately.

File: frontend/__mocks__/$app/navigation.ts (hunk 2 above)
Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts]: You are absolutely right looking at the code we directly call the callback, now i updated the file to register the callback and call it when goto is triggered, however i am not sure if that is the only thing svelte uses during navigation. please check it and let me know if you have any feedback.

File: frontend/__mocks__/$app/navigation.ts [hunk 3]
```diff
@@ -1,39 +1,68 @@
...
-    from: null,
-    to: null,
-    type: "goto",
-    willUnload: true,
-    delta: undefined,
-  });
+  afterNavigateCallback = callback;
 };
+
+export const goto =
```
Comment (dskloetd) [inline frontend/__mocks__/$app/navigation.ts:15]: Moving `goto` to the bottom makes it hard to compare it to what it was before. Can you move it back to the top?

File: frontend/__mocks__/$app/navigation.ts [hunk 4]
```diff
@@ -15,25 +15,52 @@ export const goto = async (
...
+  }
+
   page.mock({
     data: {
       universe,
       ...rest,
     },
     routeId,
   });
+  // Simulate some async operation
```
Comment (dskloetd) [inline frontend/__mocks__/$app/navigation.ts:49]: Why do we do this?

File: frontend/__mocks__/$app/navigation.ts [hunk 5]
```diff
@@ -15,25 +15,52 @@ export const goto = async (
...
 
 export const afterNavigate = (callback: (navigation: Navigation) => void) => {
-  callback({
-    from: null,
-    to: null,
-    type: "goto",
-    willUnload: true,
-    delta: undefined,
-  });
+  afterNavigateCallback = callback;
```
Comment (dskloetd) [inline frontend/__mocks__/$app/navigation.ts:65]: `afterNavigation` could be called multiple times and we'd ignore all but the last call.
Can we either not ignore them, or at least throw an error if it's called multiple times?
Also, currently if we do multiple navigations, `afterNavigateCallback` would be called multiple times, right?

Is all this logic perhaps premature? Do we actually have tests that rely on this?

File: frontend/src/lib/components/header/LinkToCanisters.svelte [hunk 6]
```diff
@@ -2,6 +2,12 @@
   import { i18n } from "$lib/stores/i18n";
   import { IconExplore } from "@dfinity/gix-components";
   import { canistersPathStore } from "$lib/derived/paths.derived";
+  import { createEventDispatcher } from "svelte";
+  import { beforeNavigate } from "$app/navigation";
+
+  const dispatcher = createEventDispatcher();
+
+  beforeNavigate(() => dispatcher("nnsLink"));
```
Comment (dskloetd) [inline frontend/src/lib/components/header/LinkToCanisters.svelte:10]: Should there be a unit test for this?
You introduced quite a bit of testing framework for this, but I can't tell if it's actually tested.

Comment (coskucinkilic) [reply frontend/src/lib/components/header/LinkToCanisters.svelte:10]: yes and when i try to actually test it it didn't work so i upgraded the navigation mock. Please review the new file and how i use it.

File: frontend/__mocks__/$app/navigation.ts (hunk 2 above)
Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts]: After navigation is not utilised anywhere but i kept it in the current update incase someone would like to use it.

File: frontend/__mocks__/$app/navigation.ts [hunk 7]
```diff
@@ -1,39 +1,106 @@
...
+
   page.mock({
     data: {
       universe,
       ...rest,
     },
     routeId,
   });
+
+  await completePromise;
```
Comment (dskloetd) [inline frontend/__mocks__/$app/navigation.ts:61]: Why is the promise created on line 25 and then awaited here?
Would it have a different effect if it's created and awaited here?

By the way, that's exactly what [runResolvedPromises](https://github.com/dfinity/nns-dapp/blob/76f74fea668616de20dbde5e72d11948ad53d9d0/frontend/src/tests/utils/timers.test-utils.ts#L10) does, with the only difference that that also works when fake timers are used. Maybe it makes sense to use that?

File: frontend/__mocks__/$app/navigation.ts [hunk 8]
```diff
@@ -1,39 +1,106 @@
...
+  for (const callback of beforeNavigateCallbacks) {
+    const beforeNav: BeforeNavigate = {
+      ...navigation,
+      cancel: () => {
+        cancelled = true;
+      },
+    };
+    callback(beforeNav);
+    if (cancelled) return;
+  }
```
Comment (dskloetd) [inline frontend/__mocks__/$app/navigation.ts:51]: Should `beforeNavigateCallbacks` be emptied after nagivation?
Or should the same `beforeNavigate` callback be called on multiple navigations?

File: frontend/__mocks__/$app/navigation.ts [hunk 9]
```diff
@@ -1,39 +1,106 @@
...
+) => {
+  afterNavigateCallbacks.push(callback);
+  return () => {
+    afterNavigateCallbacks = afterNavigateCallbacks.filter(
+      (cb) => cb !== callback
+    );
+  };
+};
+
+export const mockLinkClickEvent = (event) => {
```
Comment (dskloetd) [inline frontend/__mocks__/$app/navigation.ts:95]: clever

File: frontend/src/lib/components/header/AccountMenu.svelte [hunk 10]
```diff
@@ -34,9 +35,9 @@
 
         <ManageInternetIdentityButton />
 
-        <SettingsButton on:nnsLink={() => (visible = false)} />
+        <SettingsButton on:nnsLink={closeMenu} />
 
-        <LinkToCanisters />
+        <LinkToCanisters on:nnsLink={closeMenu} />
```
Comment (dskloetd) [inline frontend/src/lib/components/header/AccountMenu.svelte:40]: Is it possible to add a test that the menu closes after clicking on the canisters link?

File: frontend/src/tests/lib/components/header/LinkToCanisters.spec.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    expect(await po.hasIcon()).toBe(true);
+  });
+
+  it("should dispatch nnsLink event before navigation", async () => {
+    const { po, onNnsClick } = renderComponent();
+
+    await po.click();
+
+    expect(onNnsClick).toHaveBeenCalledTimes(1);
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/LinkToCanisters.spec.ts:61]: Can you expect it not to be called before the click?

File: frontend/vitest.setup.ts [hunk 12]
```diff
@@ -5,9 +5,13 @@ import "fake-indexeddb/auto";
 // jsdom does not implement TextEncoder
 // Polyfill the encoders with node
 import { TextDecoder, TextEncoder } from "util";
-import { vi } from "vitest";
+import { afterEach, beforeEach, expect, vi } from "vitest";
```
Comment (dskloetd) [inline frontend/vitest.setup.ts:8]: `afterEach, beforeEach, expect` don't seem to be used?

File: frontend/__mocks__/$app/navigation.ts (hunk 8 above)
Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts:51]: I don't know if we test multiple renders, I am also not sure how well the dom is maintained in unit tests,  if parent element let's say Layout, has a beforeNavigate, and internal component like AccountMenu also has one. If we trigger the account menu and goto another page which also render Layout, normally Layout shouldn't be rendered again as there is no change there. 

If we clear the callbacks, the beforeNavigate callback registered in Layout will be deleted, when AccountMenu triggers beforeCallback.

Maybe you have more indepth knowledge about how things are rendered and re-rendered in Svelte, if you believe that it won't be an issue, i can clean the callback after they are called.

File: frontend/__mocks__/$app/navigation.ts (hunk 7 above)
Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts:61]: Because i am also passing complete promise to beforeNavigation, i haven't implemented everything but i think you won't be able to cancel if the promise is complete, at least that is why i think they passed it there, but i didn't check the real implementation. 

How runResolvedPromises support fakeTimers? i can sure use runResolvePromises on line 25 and wait it at 61.

File: frontend/__mocks__/$app/navigation.ts (hunk 9 above)
Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts:95]: It does the job :D

File: frontend/src/lib/components/header/AccountMenu.svelte (hunk 10 above)
Comment (coskucinkilic) [reply frontend/src/lib/components/header/AccountMenu.svelte:40]: sure adding now

File: frontend/src/tests/lib/components/header/LinkToCanisters.spec.ts (hunk 11 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/LinkToCanisters.spec.ts:61]: adding now

File: frontend/vitest.setup.ts (hunk 12 above)
Comment (coskucinkilic) [reply frontend/vitest.setup.ts:8]: please check line 90, 93, 94. I generally add these kind of missing imports even-thought they are not part of my PR to fix missing or wrong stuff, also my IDE complains and i don't like seeing red stuff. Should I stop doing that?

File: frontend/__mocks__/$app/navigation.ts (hunk 7 above)
Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts:61]: When i add AccountMenu test, i realised that the promise in app/navigate mock has to be waited with runResolvedPromises in the test anyway.

I tried to use runResolvedPromises in app/navigate it didn't make a difference, i think it is due to it is being triggered through an event and cannot be waited from outside.

File: frontend/__mocks__/$app/navigation.ts [hunk 13]
```diff
@@ -1,39 +1,106 @@
...
   opts?: {
     replaceState?: boolean;
-    noscroll?: boolean;
-    keepfocus?: boolean;
+    noScroll?: boolean;
+    keepFocus?: boolean;
     /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
-    state?: any;
+    state?: Record<string, any>;
+    invalidateAll?: boolean;
```
Comment (dskloetd) [inline frontend/__mocks__/$app/navigation.ts:16]: Are these parameter changes related to the PR? They don't seem to be used. If they are not related, please don't change them in this PR. It makes it unclear what is necessary and what is unrelated.

Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts:16]: reverted

File: frontend/__mocks__/$app/navigation.ts (hunk 8 above)
Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts:51]: Ok leaving as is

File: frontend/__mocks__/$app/navigation.ts (hunk 7 above)
Comment (coskucinkilic) [reply frontend/__mocks__/$app/navigation.ts:61]: used runResolvePromises for completePromise

File: frontend/vitest.setup.ts (hunk 12 above)
Comment (coskucinkilic) [reply frontend/vitest.setup.ts:8]: reverted

**Review by dskloetd — APPROVED** [review]
